docs: ADR for course data independence from libraries

ormsbee opened the pull request:
This is an ADR around the idea that course content should not need to sync metadata from libraries during the import process or when copy and pasting.

ormsbee committed:
docs: ADR for course data independence from libraries

This is an ADR around the idea that course content should not need to
sync metadata from libraries during the import process or when copy and
pasting.

ormsbee commented:
FYI @bradenmacdonald, @connorhaugh, @feanil

kdmccormick commented on openedx/core/djangoapps/content_libraries/docs/decisions/0003-course-content-independence.rst:
```diff
@@ -0,0 +1,41 @@
+Context
+-------
+
+When a course uses content from a library via a ``LibraryContentBlock``, it creates a course-local copy of that content. It will also store certain metadata in the ModuleStore that connect the course copy with the original in the library, such as the usage key and version of the original, as well as the library-suggested values for various fields (e.g. ``num_attempts=10``).
```
```suggestion
When a course uses content from a library via a ``LibraryContentBlock``, it creates a course-local copy of the library's blocks. For each library block, it will also store certain metadata in the ModuleStore that connects the course copy with the original in the library, such as the usage key and version of the original, as well as the library-suggested values for various fields (e.g. ``num_attempts=10``).
```

on my first read-through I took "content" to mean "a set of blocks" rather than "each block". this might clarify it a bit.

kdmccormick commented on openedx/core/djangoapps/content_libraries/docs/decisions/0003-course-content-independence.rst:
```diff
@@ -0,0 +1,41 @@
+
+This metadata is not currently a part of the course export, and must be reconstructed by querying the library during the course import process. This causes a number of issues:
+
+#. The course might be imported into a site where the library does not exist, meaning that there's no way for us to figure out what the library-suggested values should be.
```
```suggestion
#. The course might be imported into a site where the library does not exist, meaning that there's no way for us to figure out what the library-suggested values should be. This becomes a problem for any block which does not override its library-suggested value: the block will fall back to the block-type-suggested value, which is often not appropriate.
```

a bit more color on this particular problem

kdmccormick commented on openedx/core/djangoapps/content_libraries/docs/decisions/0003-course-content-independence.rst:
```diff
@@ -0,0 +1,41 @@
+
+In all other cases, the course copy of the content should be self-sufficient and should not require "syncing" to a library to fill in missing metadata. In particular, the following operations should be possible to do without querying the library:
+
+* Import and Export of courses.
```
```suggestion
* Import and Export of courses within and across sites.
```

optional suggestion

kdmccormick commented on openedx/core/djangoapps/content_libraries/docs/decisions/0003-course-content-independence.rst:
```diff
@@ -0,0 +1,41 @@
+------------
+
+* We will need to serialize more metadata into the course export, probably by adding new fields to OLX.
+* This should remove a major source of bugs and confusion during the import process.
```
worth adding some more detail here?

```suggestion

Technical:

* We will need to serialize more metadata into the course export, probably by adding new fields to OLX. We will need to make this metadata available for library content operations in the XBlock runtime (ADRs to follow).
* We will need to update Import/Copy-Paste/Duplicate to use this metadata rather triggering a library sync.

User-facing:

* This should remove a major source of bugs and confusion during the import process.
* We should be able to begin preserving edits to library content during the Duplicate and Copy-Paste operations, which has been impossible as long as those operations have involved a library sync.
```

kdmccormick replied:
(EDIT: I combined the first and second technical bullet points)

kdmccormick reviewed (commented):
Looking good. Thanks for the quick writeup!

kdmccormick commented on openedx/core/djangoapps/content_libraries/docs/decisions/0003-course-content-independence.rst:
```diff
@@ -0,0 +1,41 @@
+3. Course Content should not depend on Library Content
+------------------------------------------------------
```
could you update this heading to `#####` and the others headings to `*****`? https://docs.openedx.org/en/latest/documentors/references/quick_reference_rst.html#headings

ormsbee replied:
Sure. I was just copying the other ADRs in that folder. I'll update them as well in a separate commit.

ormsbee committed:
fix: make content_libraries ADR headers match doc guidelines

https://docs.openedx.org/en/latest/documentors/references/quick_reference_rst.html#headings

ormsbee commented:
@kdmccormick: Accepted edit suggestions and changed the headers for this and other ADRs in that folder.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/docs/decisions/0003-course-content-independence.rst:
```diff
@@ -0,0 +1,48 @@
+
+Technical:
+
+* We will need to serialize more metadata into the course export, probably by adding new fields to OLX. We will need to make this metadata available for library content operations in the XBlock runtime (ADRs to follow).
```
I think we're already doing this for copy-paste, so you could do something similar.

If you copy block A then paste it as block B, then copy block B and view the context of the clipboard OLX (`/admin/content_staging/` in the django admin), you'll see that B's OLX includes something like `copied_from_block="block-v1:BradenX+TS100+23+type@problem+block@bed1cd95990a4b6b8f327494f9fe8c77"`

ormsbee replied:
😄 Yup, I remember–I was the one who asked for that feature in the first place, so that we could better track usage in analytics later on. @kdmccormick's going to put out an ADR detailing the specifics, but we're looking at a couple of new fields in this style, and maybe a `default-{field-name}` convention for handling library default values.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/docs/decisions/0003-course-content-independence.rst:
```diff
@@ -0,0 +1,48 @@
+Context
+*******
+
+When a course uses content from a library via a ``LibraryContentBlock``, it creates a course-local copy of the library's blocks. For each library block, it will also store certain metadata in the ModuleStore that connects the course copy with the original in the library, such as the usage key and version of the original, as well as the library-suggested values for various fields (e.g. ``num_attem ...
```
This describes the situation with modulestore. Would it be any different with Learning Core? (I had hoped that we'd be able to do less copying and more referencing, for example.)

ormsbee replied:
I'm not sure. I initially hated the fact that we were copying so much instead of referencing, but I've come to appreciate the decoupling this actually gave when considering something like the v1 -> v2 upgrade process, because it would mean that old content that was never going to be used again didn't need to be upgraded. I think that the use cases we have for cross-instance import/export also point to an area where copying is the better solution.

Internally, I think it makes sense to have a schema that allows for references + copies where we cross ownership/lifecycle boundaries. But those boundaries are really fuzzy to me at the moment. I'm strongly against having a 1:1 straight mapping of v1 -> v2 libraries because I think it's a terrible missed opportunity. But if not that, then what? Manually created v2 libraries that import multiple v1 libraries? An implicit, org-level shared content library maybe? I dunno.

bradenmacdonald commented:
I think this makes sense!

ormsbee commented:
@kdmccormick: Is this okay to merge now?

kdmccormick reviewed (approved):
whoops, sorry for the delay
